### DISCUSSION TITLE
Hide source code when creating component library

### DISCUSSION THREAD

**Discussion by mrados7** [callstack/react-native-builder-bob#306] [Q&A] [answered]
Hello 👋 . 
Is it possible to hide source code of components when building component library. 
Currently when I build component library using bob my source code is shown in node_modules when installed. 
Is it possible to hide `src/` directory in build or at least obfuscate code?

![image](https://user-images.githubusercontent.com/32796945/197350498-763971ce-6f6d-4aa8-a9f6-9e36e1cab2f4.png)

**Comment by satya164** [comment, marked as answer]
Just remove `src` from `files`, `react-native` entry, and disable `sourceMaps`

**Reply by mrados7** [reply]
Hey @satya164, thanks for reply. Already tried that but still gives me `src/` directory in final (published) package. But when I do build (`yarn prepare`) it doesn't show `src/` directory there. I new to this so I'm most likely doing some dumb mistake, just can't figure out.

* tsConfig.json:
```{
  "compilerOptions": {
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "esModuleInterop": true,
    "importsNotUsedAsValues": "error",
    "forceConsistentCasingInFileNames": true,
    "jsx": "react",
    "lib": ["esnext"],
    "module": "esnext",
    "moduleResolution": "node",
    "noFallthroughCasesInSwitch": true,
    "noImplicitReturns": true,
    "noImplicitUseStrict": false,
    "noStrictGenericChecks": false,
    "noUncheckedIndexedAccess": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "strict": true,
    "target": "esnext",
    "sourceMap": false
  }
}
```

* package.json:
```{
  "main": "lib/commonjs/index",
  "module": "lib/module/index",
  "types": "lib/typescript/index.d.ts",
  "react-native": "src/index",
  "source": "src/index",
  "files": [
    "lib",
    "android",
    "ios",
    "cpp",
    "*.podspec",
    "!lib/typescript/example",
    "!ios/build",
    "!android/build",
    "!android/gradle",
    "!android/gradlew",
    "!android/gradlew.bat",
    "!android/local.properties",
    "!**/__tests__",
    "!**/__fixtures__",
    "!**/__mocks__",
    "!**/.*"
  ],
}
```

* build generates:
![image](https://user-images.githubusercontent.com/32796945/197492830-7d331bd1-a882-4c22-8802-636b1c7e9b9d.png)

* after installation of published package to another project:
![image](https://user-images.githubusercontent.com/32796945/197496403-5241c7c5-56ac-410b-9d75-7e2f8f644847.png)

Basically all the directory structure is copied over. Package is published on github private repository -> https://npm.pkg.github.com/

**Reply by satya164** [reply]
compiler options are irrelevant. `npm` only cares about the `files` entry and `.npmignore`. you can check what gets published by running `npm pack` or `npm pack --dry-run`. after you remove `src` from `files`, it won't be included in the package which you can verify with the above commands.

what commands are you running to publish your package? that looks like your whole github repository which includes `.github`, `example` folders etc. which shouldn't be there in the published package.

though I'd also emphasize that unless you're obfuscating your code, there isn't really much hidden by not publishing `src` folder. your client-side libraries shouldn't include any secrets that you don't want others to be able to find. client-side code is always public. someone can still figure out stuff after obfuscating.

**Reply by mrados7** [reply]
Alright it was dumb mistake in the end. My github actions were using `yarn publish` which I think (correct me if wrong) doesn't care about "files" in package.json.
 I don't include any secret or something like that, but it doesn't make sense to publish source code as well when not needed. Thank you so much for your help @satya164 !

**Reply by satya164** [reply]
Yarn should work too but it has some strange behavior so I recommend using `npm` for publishing anyway

**Comment by hashamVeroke** [comment]
@satya164 need your help

While creating a React Native library using TurboModules, I want to exclude the src folder from the final npm package.

My TypeScript code lives in src/

I transpile to lib/ or dist/ before publishing

When I remove src from the published package (via .npmignore or files field in package.json), I face the following issue when installing the library in another React Native project:

**Error: ENOENT: no such file or directory, lstat '.../node_modules/mylibrary/src'
[Codegen] Done.**

This happens because React Native Codegen (triggered during pod install) is still trying to read the src folder, even though only lib/ should be required at runtime.

Expected Behavior

I should be able to ship only compiled files (lib/, dist/) in my npm package.

src folder should not be required in consuming apps.

Codegen should work with compiled output instead of raw src.

Actual Behavior

Without src, pod install fails with the above error.

This makes it impossible to hide implementation details when distributing the library.

Steps to Reproduce

Create a TurboModule library with TypeScript sources in src/.

Add a build step (tsc or Babel) to generate lib/.

Exclude src from package.json "files" or .npmignore.

Pack and install the library in another RN app.

Run pod install → ❌ Error: ENOENT: .../node_modules/mylibrary/src.

**Reply by satya164** [reply]
The codegen specs need to be published to npm. If you don't want to publish `src` then you need to move the codegen specs to a different folder and adjust the path in `codegenConfig`.
